File: sdk/go/workflow_authoring.go
package gestalt

import (
	"fmt"
	"strings"
)

// WorkflowRefInput references a run-input path in workflow authoring builders.
func WorkflowRefInput(path string) WorkflowValue {
	return WorkflowValue{Input: strings.TrimSpace(path)}
}

// WorkflowRefSignal references an activation signal path in workflow authoring builders.
func WorkflowRefSignal(path string) WorkflowValue {
	return WorkflowValue{Signal: strings.TrimSpace(path)}
}

// WorkflowRefStepOutput references a prior step output path.
func WorkflowRefStepOutput(stepID, path string) WorkflowValue {
	return WorkflowValue{StepOutput: &WorkflowStepOutputSource{
		StepID: strings.TrimSpace(stepID),
		Path:   strings.TrimSpace(path),
	}}
}

// WorkflowRefStepInput references a prior step input path.
func WorkflowRefStepInput(stepID, path string) WorkflowValue {
	return WorkflowValue{StepInput: &WorkflowStepInputSource{
		StepID: strings.TrimSpace(stepID),
		Path:   strings.TrimSpace(path),
	}}
}

// WorkflowRefLiteral builds a literal workflow value.
func WorkflowRefLiteral(value any) WorkflowValue {
	return WorkflowValue{Literal: value, LiteralSet: true}
}

// WorkflowRefTemplate builds a template workflow value.
func WorkflowRefTemplate(template string) WorkflowValue {
	text := WorkflowText{Template: template}
	return WorkflowValue{Template: &text}
}

// WorkflowRefObject builds an object workflow value from field builders.
func WorkflowRefObject(fields map[string]WorkflowValue) WorkflowValue {
	return WorkflowValue{Object: fields}
}

// WorkflowRefArray builds an array workflow value.
func WorkflowRefArray(values ...WorkflowValue) WorkflowValue {
	return WorkflowValue{Array: values}
}

// DefineWorkflowOptions configures a typed workflow builder.
type DefineWorkflowOptions struct {
	ID     string
	RunAs  string
	Paused bool
}

// WorkflowEventActivationOptions configures an event activation.
type WorkflowEventActivationOptions struct {
	ID      string
	Source  string
	Subject string
	Paused  bool
}

// WorkflowScheduleActivationOptions configures a schedule activation.
type WorkflowScheduleActivationOptions struct {
	ID       string
	Timezone string
	Paused   bool
}

// WorkflowEventActivationConfig is accepted by WorkflowBuilder.On.
type WorkflowEventActivationConfig struct {
	Type     string
	MapInput func(WorkflowEventScope) map[string]WorkflowValue
	Options  WorkflowEventActivationOptions
}

// WorkflowScheduleActivationConfig is accepted by WorkflowBuilder.On.
type WorkflowScheduleActivationConfig struct {
	Cron     string
	MapInput func(WorkflowActivationScope) map[string]WorkflowValue
	Options  WorkflowScheduleActivationOptions
}

// WorkflowEventScope exposes signal references for event activation mapping.
type WorkflowEventScope struct{}

// Data returns a signal reference under event.data.
func (WorkflowEventScope) Data(path string) WorkflowValue {
	return WorkflowRefSignal(joinWorkflowPath("data", path))
}

// Field returns a signal reference at the given path.
func (WorkflowEventScope) Field(path string) WorkflowValue {
	return WorkflowRefSignal(path)
}

// WorkflowActivationScope exposes input references for schedule activation mapping.
type WorkflowActivationScope struct{}

// Input returns a run-input reference.
func (WorkflowActivationScope) Input(path string) WorkflowValue {
	return WorkflowRefInput(path)
}

// WorkflowStepScope exposes references available while authoring a step.
type WorkflowStepScope struct{}

// Input returns a run-input reference.
func (WorkflowStepScope) Input(path string) WorkflowValue {
	return WorkflowRefInput(path)
}

// Signal returns a signal reference.
func (WorkflowStepScope) Signal(path string) WorkflowValue {
	return WorkflowRefSignal(path)
}

// StepOutput returns a prior step output reference.
func (WorkflowStepScope) StepOutput(stepID, path string) WorkflowValue {
	return WorkflowRefStepOutput(stepID, path)
}

// StepInput returns a prior step input reference.
func (WorkflowStepScope) StepInput(stepID, path string) WorkflowValue {
	return WorkflowRefStepInput(stepID, path)
}

// WorkflowStepAppConfig configures an app step action.
type WorkflowStepAppConfig struct {
	Name           string
	Operation      string
	Input          func(WorkflowStepScope) map[string]WorkflowValue
	Connection     string
	Instance       string
	CredentialMode string
}

// WorkflowStepAgentMessageConfig configures one agent message.
type WorkflowStepAgentMessageConfig struct {
	Role string
	Text string
}

// WorkflowStepAgentConfig configures an agent step action.
type WorkflowStepAgentConfig struct {
	Provider     string
	Model        string
	SessionKey   string
	Prompt       string
	Messages     []WorkflowStepAgentMessageConfig
	Tools        []AgentToolRef
	Output       *AgentOutput
	ModelOptions any
}

// WorkflowStepWhenConfig configures a step guard.
type WorkflowStepWhenConfig struct {
	Value  WorkflowValue
	Equals any
}

// WorkflowStepConfig configures one workflow step.
type WorkflowStepConfig struct {
	Inputs         func(WorkflowStepScope) map[string]WorkflowValue
	App            *WorkflowStepAppConfig
	Agent          *WorkflowStepAgentConfig
	When           *WorkflowStepWhenConfig
	TimeoutSeconds int32
	Metadata       any
}

// WorkflowBuilder incrementally authors a workflow definition spec.
type WorkflowBuilder struct {
	id          string
	runAs       string
	paused      bool
	activations []WorkflowActivation
	steps       []WorkflowStep
}

// Event creates an event activation configuration.
func Event(typeName string, mapInput func(WorkflowEventScope) map[string]WorkflowValue, options WorkflowEventActivationOptions) WorkflowEventActivationConfig {
	return WorkflowEventActivationConfig{
		Type:     typeName,
		MapInput: mapInput,
		Options:  options,
	}
}

// Schedule creates a schedule activation configuration.
func Schedule(cron string, mapInput func(WorkflowActivationScope) map[string]WorkflowValue, options WorkflowScheduleActivationOptions) WorkflowScheduleActivationConfig {
	return WorkflowScheduleActivationConfig{
		Cron:     cron,
		MapInput: mapInput,
		Options:  options,
	}
}

// DefineWorkflow starts a fluent workflow definition builder.
func DefineWorkflow(options DefineWorkflowOptions) (*WorkflowBuilder, error) {
	runAs := strings.TrimSpace(options.RunAs)
	if runAs == "" {
		return nil, fmt.Errorf("DefineWorkflow requires RunAs")
	}
	id := strings.TrimSpace(options.ID)
	if id == "" {
		return nil, fmt.Errorf("DefineWorkflow requires ID")
	}
	return &WorkflowBuilder{
		id:     id,
		runAs:  runAs,
		paused: options.Paused,
	}, nil
}

// On appends one activation to the builder.
func (b *WorkflowBuilder) On(activation any) *WorkflowBuilder {
	switch typed := activation.(type) {
	case WorkflowEventActivationConfig:
		activationID := strings.TrimSpace(typed.Options.ID)
		if activationID == "" {
			activationID = typed.Type
		}
		var input WorkflowValue
		if typed.MapInput != nil {
			input = WorkflowValue{Object: typed.MapInput(WorkflowEventScope{})}
		}
		b.activations = append(b.activations, WorkflowActivation{
			ID:     activationID,
			Paused: typed.Options.Paused,
			Event: &WorkflowEventActivation{Match: &WorkflowEventMatch{
				Type:    typed.Type,
				Source:  typed.Options.Source,
				Subject: typed.Options.Subject,
			}},
			Input: input,
		})
	case WorkflowScheduleActivationConfig:
		activationID := strings.TrimSpace(typed.Options.ID)
		if activationID == "" {
			activationID = typed.Cron
		}
		var input WorkflowValue
		if typed.MapInput != nil {
			input = WorkflowValue{Object: typed.MapInput(WorkflowActivationScope{})}
		}
		b.activations = append(b.activations, WorkflowActivation{
			ID:     activationID,
			Paused: typed.Options.Paused,
			Schedule: &WorkflowScheduleActivation{
				Cron:     typed.Cron,
				Timezone: typed.Options.Timezone,
			},
			Input: input,
		})
	}
	return b
}

// Step appends one step to the builder.
func (b *WorkflowBuilder) Step(stepID string, config WorkflowStepConfig) *WorkflowBuilder {
	scope := WorkflowStepScope{}
	step := WorkflowStep{ID: stepID}
	if config.Inputs != nil {
		step.Inputs = config.Inputs(scope)
	}
	if config.App != nil {
		var input WorkflowValue
		if config.App.Input != nil {
			input = WorkflowValue{Object: config.App.Input(scope)}
		}
		step.App = &WorkflowStepAppCall{
			Name:           config.App.Name,
			Operation:      config.App.Operation,
			Input:          input,
			Connection:     config.App.Connection,
			Instance:       config.App.Instance,
			CredentialMode: config.App.CredentialMode,
		}
	}
	if config.Agent != nil {
		messages := make([]WorkflowAgentMessage, 0, len(config.Agent.Messages))
		for _, message := range config.Agent.Messages {
			messages = append(messages, WorkflowAgentMessage{
				Role: message.Role,
				Text: WorkflowText{Template: message.Text},
			})
		}
		step.Agent = &WorkflowStepAgentTurn{
			Provider:     config.Agent.Provider,
			Model:        config.Agent.Model,
			SessionKey:   config.Agent.SessionKey,
			Prompt:       WorkflowText{Template: config.Agent.Prompt},
			Messages:     messages,
			Tools:        config.Agent.Tools,
			Output:       config.Agent.Output,
			ModelOptions: config.Agent.ModelOptions,
		}
	}
	if config.When != nil {
		step.When = &WorkflowStepWhen{
			Value:  config.When.Value,
			Equals: config.When.Equals,
		}
	}
	step.TimeoutSeconds = config.TimeoutSeconds
	step.Metadata = config.Metadata
	b.steps = append(b.steps, step)
	return b
}

// ToSpec returns the authored workflow definition spec.
func (b *WorkflowBuilder) ToSpec() WorkflowDefinitionSpec {
	var target *BoundWorkflowTarget
	if len(b.steps) > 0 {
		target = &BoundWorkflowTarget{Steps: append([]WorkflowStep(nil), b.steps...)}
	}
	return WorkflowDefinitionSpec{
		ID:          b.id,
		RunAs:       &Subject{ID: b.runAs},
		Paused:      b.paused,
		Activations: append([]WorkflowActivation(nil), b.activations...),
		Target:      target,
	}
}

// WorkflowComposeText builds workflow template text from literals and workflow value references.
func WorkflowComposeText(parts ...any) string {
	var builder strings.Builder
	for _, part := range parts {
		switch typed := part.(type) {
		case string:
			builder.WriteString(typed)
		case WorkflowValue:
			builder.WriteString(workflowValueToTemplatePlaceholder(typed))
		default:
			panic(fmt.Sprintf("WorkflowComposeText parts must be strings or WorkflowValue, got %T", part))
		}
	}
	return builder.String()
}

func workflowValueToTemplatePlaceholder(value WorkflowValue) string {
	if value.Input != "" {
		return fmt.Sprintf("${{ input.%s }}", strings.TrimSpace(value.Input))
	}
	if value.Signal != "" {
		return fmt.Sprintf("${{ signal.%s }}", strings.TrimSpace(value.Signal))
	}
	if value.StepOutput != nil {
		return fmt.Sprintf(
			"${{ steps.%s.outputs.%s }}",
			strings.TrimSpace(value.StepOutput.StepID),
			strings.TrimSpace(value.StepOutput.Path),
		)
	}
	if value.StepInput != nil {
		return fmt.Sprintf(
			"${{ steps.%s.inputs.%s }}",
			strings.TrimSpace(value.StepInput.StepID),
			strings.TrimSpace(value.StepInput.Path),
		)
	}
	panic("WorkflowComposeText references must be input, signal, step output, or step input paths")
}

// ResolveWorkflowDefinitionSpec accepts either a builder or a raw spec.
func ResolveWorkflowDefinitionSpec(input any) (WorkflowDefinitionSpec, error) {
	switch typed := input.(type) {
	case *WorkflowBuilder:
		return typed.ToSpec(), nil
	case WorkflowBuilder:
		return typed.ToSpec(), nil
	case WorkflowDefinitionSpec:
		return typed, nil
	case *WorkflowDefinitionSpec:
		if typed == nil {
			return WorkflowDefinitionSpec{}, nil
		}
		return *typed, nil
	default:
		return WorkflowDefinitionSpec{}, fmt.Errorf("unsupported workflow definition spec input %T", input)
	}
}

File: sdk/rust/src/workflow_authoring.rs
//! Typed workflow-definition authoring helpers.
#![allow(missing_docs)]

use std::collections::BTreeMap;

use serde_json::{Map, Value};

use crate::agent::AgentOutput;
use crate::app::AgentToolRef;
use crate::workflow::{
    BoundWorkflowTarget, WorkflowActivation, WorkflowActivationTrigger, WorkflowAgentMessage,
    WorkflowDefinitionSpec, WorkflowEventActivation, WorkflowEventMatch,
    WorkflowScheduleActivation, WorkflowStep, WorkflowStepAction, WorkflowStepAgentTurn,
    WorkflowStepAppCall, WorkflowStepWhen, WorkflowText, WorkflowValue, WorkflowValueKind,
};

type WorkflowMap = BTreeMap<String, WorkflowValue>;
type WorkflowMapFn = Box<dyn Fn() -> WorkflowMap + Send + Sync>;

fn value(kind: WorkflowValueKind) -> WorkflowValue {
    WorkflowValue { kind: Some(kind) }
}

/// References a run-input path.
pub fn input(path: impl Into<String>) -> WorkflowValue {
    value(WorkflowValueKind::Input(
        crate::workflow::WorkflowPathSource { path: path.into() },
    ))
}

/// References an activation signal path.
pub fn signal(path: impl Into<String>) -> WorkflowValue {
    value(WorkflowValueKind::Signal(
        crate::workflow::WorkflowPathSource { path: path.into() },
    ))
}

/// References a prior step output path.
pub fn step_output(step_id: impl Into<String>, path: impl Into<String>) -> WorkflowValue {
    value(WorkflowValueKind::StepOutput(
        crate::workflow::WorkflowStepOutputSource {
            step_id: step_id.into(),
            path: path.into(),
        },
    ))
}

/// References a prior step input path.
pub fn step_input(step_id: impl Into<String>, path: impl Into<String>) -> WorkflowValue {
    value(WorkflowValueKind::StepInput(
        crate::workflow::WorkflowStepInputSource {
            step_id: step_id.into(),
            path: path.into(),
        },
    ))
}

/// Builds a literal workflow value.
pub fn literal(literal: Value) -> crate::Result<WorkflowValue> {
    Ok(value(WorkflowValueKind::Literal(literal)))
}

/// Builds a workflow template value.
pub fn template(template: impl Into<String>) -> WorkflowValue {
    value(WorkflowValueKind::Template(WorkflowText {
        template: template.into(),
    }))
}

/// Builds an object workflow value.
pub fn object(fields: WorkflowMap) -> WorkflowValue {
    value(WorkflowValueKind::Object(crate::workflow::WorkflowObject {
        fields,
    }))
}

/// Builds an array workflow value.
pub fn array(values: Vec<WorkflowValue>) -> WorkflowValue {
    value(WorkflowValueKind::Array(crate::workflow::WorkflowArray {
        values,
    }))
}

#[derive(Clone, Debug)]
pub struct WorkflowBuilder {
    id: String,
    run_as: String,
    paused: bool,
    activations: Vec<WorkflowActivation>,
    steps: Vec<WorkflowStep>,
}

/// Starts a fluent workflow definition builder.
pub fn define_workflow(
    id: impl Into<String>,
    run_as: impl Into<String>,
) -> crate::Result<WorkflowBuilder> {
    let id = id.into();
    let run_as = run_as.into();
    if id.trim().is_empty() {
        return Err(crate::Error::bad_request("define_workflow requires id"));
    }
    if run_as.trim().is_empty() {
        return Err(crate::Error::bad_request("define_workflow requires run_as"));
    }
    Ok(WorkflowBuilder {
        id,
        run_as,
        paused: false,
        activations: Vec::new(),
        steps: Vec::new(),
    })
}

impl WorkflowBuilder {
    /// Sets whether the definition starts paused.
    pub fn paused(mut self, paused: bool) -> Self {
        self.paused = paused;
        self
    }

    pub fn on(mut self, activation: WorkflowActivationConfig) -> Self {
        let (id, paused, trigger, input) = match activation.kind {
            WorkflowActivationKind::Event {
                type_name,
                map_input,
            } => (
                type_name.clone(),
                false,
                WorkflowActivationTrigger::Event(WorkflowEventActivation {
                    r#match: Some(WorkflowEventMatch {
                        r#type: type_name,
                        ..Default::default()
                    }),
                }),
                Some(object(map_input())),
            ),
            WorkflowActivationKind::Schedule { cron, map_input } => (
                cron.clone(),
                false,
                WorkflowActivationTrigger::Schedule(WorkflowScheduleActivation {
                    cron,
                    ..Default::default()
                }),
                Some(object(map_input())),
            ),
        };
        self.activations.push(WorkflowActivation {
            id,
            paused,
            trigger: Some(trigger),
            input,
        });
        self
    }

    pub fn step(mut self, step_id: impl Into<String>, config: WorkflowStepConfig) -> Self {
        assert!(
            !(config.app.is_some() && config.agent.is_some()),
            "workflow step cannot configure both app and agent actions"
        );
        let mut step = WorkflowStep {
            id: step_id.into(),
            ..Default::default()
        };
        if let Some(inputs) = config.inputs {
            step.inputs = inputs();
        }
        if let Some(app) = config.app {
            step.action = Some(WorkflowStepAction::App(WorkflowStepAppCall {
                name: app.name,
                operation: app.operation,
                input: app.input.map(|callback| object(callback())),
                connection: app.connection,
                instance: app.instance,
                credential_mode: app.credential_mode,
            }));
        }
        if let Some(agent) = config.agent {
            step.action = Some(WorkflowStepAction::Agent(WorkflowStepAgentTurn {
                provider: agent.provider,
                model: agent.model,
                session_key: agent.session_key,
                prompt: Some(agent.prompt),
                messages: agent
                    .messages
                    .into_iter()
                    .map(|message| WorkflowAgentMessage {
                        role: message.role,
                        text: Some(WorkflowText {
                            template: message.text,
                        }),
                        ..Default::default()
                    })
                    .collect(),
                tools: agent.tools,
                output: agent.output,
                model_options: agent.model_options,
            }));
        }
        step.when = config.when.map(|when| WorkflowStepWhen {
            value: Some(when.value),
            equals: when.equals,
        });
        step.timeout_seconds = config.timeout_seconds;
        step.metadata = config.metadata;
        self.steps.push(step);
        self
    }

    /// Consumes the builder and returns the generated workflow definition spec.
    pub fn to_spec(self) -> WorkflowDefinitionSpec {
        WorkflowDefinitionSpec {
            id: self.id,
            run_as: self.run_as,
            paused: self.paused,
            activations: self.activations,
            target: (!self.steps.is_empty()).then_some(BoundWorkflowTarget { steps: self.steps }),
        }
    }
}

pub struct WorkflowActivationConfig {
    kind: WorkflowActivationKind,
}

enum WorkflowActivationKind {
    Event {
        type_name: String,
        map_input: WorkflowMapFn,
    },
    Schedule {
        cron: String,
        map_input: WorkflowMapFn,
    },
}

/// Creates an event activation for [`WorkflowBuilder::on`].
pub fn event<F>(type_name: impl Into<String>, map_input: F) -> WorkflowActivationConfig
where
    F: Fn() -> BTreeMap<String, WorkflowValue> + Send + Sync + 'static,
{
    WorkflowActivationConfig {
        kind: WorkflowActivationKind::Event {
            type_name: type_name.into(),
            map_input: Box::new(map_input),
        },
    }
}

/// Creates a schedule activation for [`WorkflowBuilder::on`].
pub fn schedule<F>(cron: impl Into<String>, map_input: F) -> WorkflowActivationConfig
where
    F: Fn() -> BTreeMap<String, WorkflowValue> + Send + Sync + 'static,
{
    WorkflowActivationConfig {
        kind: WorkflowActivationKind::Schedule {
            cron: cron.into(),
            map_input: Box::new(map_input),
        },
    }
}

pub struct WorkflowStepAppConfig {
    pub name: String,
    pub operation: String,
    input: Option<WorkflowMapFn>,
    pub connection: String,
    pub instance: String,
    pub credential_mode: String,
}

impl WorkflowStepAppConfig {
    /// Creates an app step configuration. Add mapped input with [`Self::with_input`].
    pub fn new(name: impl Into<String>, operation: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            operation: operation.into(),
            input: None,
            connection: String::new(),
            instance: String::new(),
            credential_mode: String::new(),
        }
    }

    /// Sets a zero-argument input mapper; the closure is boxed by the SDK.
    pub fn with_input<F>(mut self, input: F) -> Self
    where
        F: Fn() -> BTreeMap<String, WorkflowValue> + Send + Sync + 'static,
    {
        self.input = Some(Box::new(input));
        self
    }
}

impl Default for WorkflowStepAppConfig {
    fn default() -> Self {
        Self::new("", "")
    }
}

impl WorkflowStepAppConfig {
    pub fn with_connection(mut self, connection: impl Into<String>) -> Self {
        self.connection = connection.into();
        self
    }

    pub fn with_instance(mut self, instance: impl Into<String>) -> Self {
        self.instance = instance.into();
        self
    }

    pub fn with_credential_mode(mut self, credential_mode: impl Into<String>) -> Self {
        self.credential_mode = credential_mode.into();
        self
    }
}

#[derive(Clone)]
pub struct WorkflowStepAgentMessageConfig {
    pub role: String,
    pub text: String,
}

#[derive(Clone, Default)]
pub struct WorkflowStepAgentConfig {
    pub provider: String,
    pub model: String,
    pub session_key: String,
    pub prompt: WorkflowText,
    pub messages: Vec<WorkflowStepAgentMessageConfig>,
    pub tools: Vec<AgentToolRef>,
    pub output: Option<AgentOutput>,
    pub model_options: Option<Map<String, Value>>,
}

impl WorkflowStepAgentConfig {
    pub fn new(provider: impl Into<String>) -> Self {
        Self {
            provider: provider.into(),
            ..Default::default()
        }
    }

    pub fn with_prompt(mut self, prompt: WorkflowText) -> Self {
        self.prompt = prompt;
        self
    }
}

#[derive(Clone)]
pub struct WorkflowStepWhenConfig {
    pub value: WorkflowValue,
    pub equals: Option<Value>,
}

#[derive(Default)]
pub struct WorkflowStepConfig {
    inputs: Option<WorkflowMapFn>,
    pub app: Option<WorkflowStepAppConfig>,
    pub agent: Option<WorkflowStepAgentConfig>,
    pub when: Option<WorkflowStepWhenConfig>,
    pub timeout_seconds: i32,
    pub metadata: Option<Map<String, Value>>,
}

impl WorkflowStepConfig {
    pub fn with_inputs<F>(mut self, inputs: F) -> Self
    where
        F: Fn() -> BTreeMap<String, WorkflowValue> + Send + Sync + 'static,
    {
        self.inputs = Some(Box::new(inputs));
        self
    }
}

fn value_placeholder(value: &WorkflowValue) -> crate::Result<String> {
    let result = match value.kind.as_ref() {
        Some(WorkflowValueKind::Input(source)) => format!("${{{{ input.{} }}}}", source.path),
        Some(WorkflowValueKind::Signal(source)) => format!("${{{{ signal.{} }}}}", source.path),
        Some(WorkflowValueKind::StepOutput(source)) => {
            format!(
                "${{{{ steps.{}.outputs.{} }}}}",
                source.step_id, source.path
            )
        }
        Some(WorkflowValueKind::StepInput(source)) => {
            format!("${{{{ steps.{}.inputs.{} }}}}", source.step_id, source.path)
        }
        _ => {
            return Err(crate::Error::bad_request(
                "text references must be path values",
            ));
        }
    };
    Ok(result)
}

/// Composes workflow prompt text from workflow value references.
pub fn text(parts: &[WorkflowValue]) -> crate::Result<WorkflowText> {
    let mut template = String::new();
    for part in parts {
        template.push_str(&value_placeholder(part)?);
    }
    Ok(WorkflowText { template })
}

pub enum WorkflowDefinitionSpecOrBuilder {
    Spec(WorkflowDefinitionSpec),
    Builder(WorkflowBuilder),
}

impl From<WorkflowBuilder> for WorkflowDefinitionSpecOrBuilder {
    fn from(builder: WorkflowBuilder) -> Self {
        Self::Builder(builder)
    }
}

impl From<WorkflowDefinitionSpec> for WorkflowDefinitionSpecOrBuilder {
    fn from(spec: WorkflowDefinitionSpec) -> Self {
        Self::Spec(spec)
    }
}

impl WorkflowDefinitionSpecOrBuilder {
    fn into_spec(self) -> WorkflowDefinitionSpec {
        match self {
            Self::Spec(spec) => spec,
            Self::Builder(builder) => builder.to_spec(),
        }
    }
}

/// Applies a workflow definition from a raw spec or an authored builder.
pub async fn apply_workflow_definition(
    workflow: &mut crate::workflow::Workflow,
    provider: String,
    idempotency_key: String,
    spec: Option<impl Into<WorkflowDefinitionSpecOrBuilder>>,
) -> Result<crate::workflow::WorkflowDefinition, crate::rpc_support::GestaltError> {
    workflow
        .apply_definition(
            provider,
            idempotency_key,
            spec.map(|value| value.into().into_spec()),
        )
        .await
}
